multiuser/machine-auth/che-multiuser-machine-authentication/src/main/java/org/eclipse/che/multiuser/machine/authentication/server/signature/SignatureKeyManager.java

```diff
@@ -57,56 +60,79 @@
 
   private final String algorithm;
   private final SignatureKeyDao signatureKeyDao;
+  private final EventService eventService;
+  private final EventSubscriber<?> workspaceEventsSubscriber;
 
   @Inject
   @SuppressWarnings("unused")
   private DBInitializer dbInitializer;
 
-  private KeyPair cachedPair;
+  private Map<String, KeyPair> cachedPair = new HashMap<>();
 
   @Inject
   public SignatureKeyManager(
       @Named("che.auth.signature_key_size") int keySize,
       @Named("che.auth.signature_key_algorithm") String algorithm,
+      EventService eventService,
       SignatureKeyDao signatureKeyDao) {
     this.keySize = keySize;
     this.algorithm = algorithm;
+    this.eventService = eventService;
     this.signatureKeyDao = signatureKeyDao;
+    this.workspaceEventsSubscriber =
+        new EventSubscriber<WorkspaceStatusEvent>() {
+          @Override
+          public void onEvent(WorkspaceStatusEvent event) {
+            switch (event.getStatus()) {
+              case STOPPED:
+                try {
+                  cachedPair.remove(event.getWorkspaceId());
+                  signatureKeyDao.remove(event.getWorkspaceId());
+                } catch (ServerException e) {
+                  LOG.error(e.getLocalizedMessage(), e);
```

Comment:
Please add context message here. Also it would be better not to use localized message here